servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -5961,11 +5962,17 @@ List<ServiceIdentity> setupServiceIdentityList(AthenzDomain domain, Boolean publ
             }
         }
 
+        if (tagKey != null) {
+            services = services.stream()
+                    .filter(service -> filterByTag(tagKey, tagValue, service, ServiceIdentity::getTags))
+                    .collect(Collectors.toList());
+        }
+
         return services;
     }
 
     public ServiceIdentities getServiceIdentities(ResourceContext ctx, String domainName,
-            Boolean publicKeys, Boolean hosts) {
+            Boolean publicKeys, Boolean hosts, String tagKey, String tagValue) {
```

Comment:
We should have validate calls for tagKey and tagValue values to make sure they're CompoundNames if they're specified.

Reply:
I didn't understand this comment can you please explain? 
this is the same way that we do it for roles and groups.

Reply:
looks like we missed in other places as well. as part of any input data, we must always validate the data. we can't allow any data unchecked from the user to follow through the system if we have specific constrains set in our rdl. So in this case both are defined as CompoundName but could also be null. So we need to validate the data with:

```
if (!StringUtils.isEmpty(tagKey) {
          validate(tagKey, TYPE_COMPOUND_NAME, caller);
}
if (!StringUtils.isEmpty(tagValue) {
          validate(tagValue, TYPE_COMPOUND_NAME, caller);
}
```